code-block/src/components/CodeEditor/CodeEditor.tsx

```diff
@@ -27,21 +27,29 @@ export const CodeEditor: FunctionComponent<{
     setContent((content) => ({
       ...content,
       code: value,
-      highlightStates: withLength(
-        content.highlightStates,
-        lineCount,
-        defaultHighlightStateOption.value,
-      ),
+      highlightStates: options.highlightStates
+        ? withLength(
+            content.highlightStates ?? [],
+            lineCount,
+            defaultHighlightStateOption.value,
+          )
+        : undefined,
     }))
 
-  const handleLineNumberClick = (line: number) =>
+  const handleLineNumberClick = (line: number, lineCount: number) =>
     setContent((content) => ({
       ...content,
-      highlightStates: toggleLine(
-        options.highlightStates,
-        content.highlightStates,
-        line,
-      ),
+      highlightStates: options.highlightStates
+        ? toggleLine(
+            options.highlightStates,
+            withLength(
+              content.highlightStates ?? [],
+              lineCount,
+              defaultHighlightStateOption.value,
+            ),
+            line,
+          )
+        : undefined,
```

Comment:
Maybe I should write some unit tests for this section?

Reply:
I answered my own question: _yes_. 

- I extracted the updater functions here: 9e27fe87aaba13115d35114bc623077bd120f6a4
- I added the tests here: 50888abe21a5b8e13d4bbdb86a3fe1fe8289ab5c